source/kraken/readme.md

```diff
@@ -0,0 +1,201 @@
+4. build raptor
+5. build relations
+6. build proximitylist
+
+When a data loading occurs after startup the process is the same but is done on another dataset, this means that
+memory usage will double during reload.
+There is no service interruption as we have two datasets in memory, there is no locking done to prevent blocking
+requests. Swap of dataset is done by an atomic swap of pointer.
+
+## Realtime integration
+
+In this chapter, 'realtime' means any modification of the static data, hence disruptions from Chaos or
+realtime_update from Kirin.
+
+These data are received from rabbitmq, kraken will handle multiple messages in batch (up to 5000).
+The process is very similar to a data reload, it is done in background on another dataset to prevent impacts
+on requests.
+
+The following actions are done:
+1. clone `Data` to have a writable dataset, this is quite slow
+2. update `Data` with the realtime data
+3. build relations
+4. build autocomplete
+5. build raptor
+6. build proximitylist
+7. rebuild raptor cache from the previous Data
+8. switch `Data`
+
+Rebuilding raptor's cache is not strictly required, but it reducse the slowdown of the first few requests on the new
+dataset.
+
+## Request handling
+The main thread execute the [`ZMQ LoadBalancer`](https://github.com/CanalTP/utils/blob/master/zmq.h) that
+dispatch requests to available worker threads, it only forward the requests to a worker and respond to the client
+once the worker have finished. There is no serialization done here, only (unneeded) copy of bytes.
+Wait queue is not handled by the load balancer, it won't accept a request if there is no worker available, the
+queue is managed by zmq.
+Communication between threads is done with [zmq inproc sockets](http://api.zeromq.org/2-1:zmq-inproc).
+
+Workers threads start by registering themselves to the load balancer and start waiting for requests. Each threads
+can only handle one request concurrently.
+
+When a request is received the following actions occur:
+1. read request and deserialize it
+2. if enabled create a deadline object to terminate requests
+3. acquire a readonly Data object that will be used for this request
+4. check if deadline has expired
+5. initialize worker if needed: first request with a new dataset
+    1. creation of raptor planner
+    2. creation of streetnetwork planner
+6. initialisation of `pb_creator` used to build the response
+7. execute handler of this API
+8. build response from `pb_creator`
+9. respond to LB
+
+In case of error the thread will respond with the error message, termination by deadline is handled like an
+error.
+
+### Raptor cache
+The raptor cache is a structure shared between every raptor planner that contains an optimized
+reprensation of stoptimes for a specific period of time. Every request that compute stoptimes (journeys,
+departures, *_schedules) will need a raptor cache for the date of the request as raptor cache only contains 48
```

Comment:
To me, the 48h are kind of side-consideration, the most important is that a raptor cache is attached and usable on one precise day, no more no less.
Maybe this is clearer:
```suggestion
departures, *_schedules) will use the raptor cache attached the starting day of the request (raptor cache only contains 48 hours of data to handle pass-midnight journeys)
```